### ISSUE TITLE
Support baselines

### ISSUE THREAD

**Issue by edgarrmondragon** [charliermarsh/ruff#1149]
Baselines would allow developers to introduce Ruff to their codebase incrementally by ignoring existing errors and address them as they reappear when making changes to the code.

Supported by some tools in the flake8 ecosystem:

- Bandit: https://bandit.readthedocs.io/en/latest/start.html#baseline
- Flakeheaven: https://wemake-python-styleguide.readthedocs.io/en/latest/pages/usage/integrations/flakeheaven.html#legacy-first-1

A very basic proposal for this in ruff:

### In the CLI

```
ruff . --exit-zero --format baseline > .ruff_baseline
ruff . --baseline .ruff_baseline
```

### In pyproject.toml

```toml
[tool.ruff]
baseline = ".ruff_baseline"
```

**Comment by charliermarsh** [comment]
That's interesting! I've never seen this before. I think we could support it? I'd been hoping that `--add-noqa` would serve this purpose. What do you see as the strengths of a baseline file vis-a-vis that workflow?

One point of confusion for me w/r/t the baseline file is that if it's just storing errors + line numbers, then if you edit a file (e.g., add a function to the top of a file that has some ignored errors), doesn't the baseline get invalidated? Since all the errors are pushed to new lines?

**Comment by jfmengels** [comment]
I would recommend looking into what [`elm-review`](https://elm-review.com/) did, with its error suppression system. I described the system in this [blog post](https://jfmengels.net/stop-the-bleed/). The baseline doesn't include the line numbers — but instead counts the number of problems per file and per rule — meaning it doesn't suffer from the problem @charliermarsh was rightfully wary of.

My experience with this system is very positive. I would definitely recommend it over adding a bunch of noqa comments everywhere, because now it becomes unclear what purposefully got a `noqa` and what got one because someone used `--add-noqa` because there were too many errors. It's also a better system than turning the rules on as "warnings" (which ruff doesn't support but IMO shouldn't) which some people tend to do when there are too many existing errors.

FYI, `elm-review` is a very beloved linter for the Elm language, and it supports neither `noqa` comments nor severity levels (warning/error), only ignored files through the configuration and this suppression/baseline system. And it's working really well.
Some more information around this in its [documentation](https://package.elm-lang.org/packages/jfmengels/elm-review/latest/#is-there-a-way-to-ignore-errors-)

Happy to answer any questions on the topic (and linters in general 😁)

**Comment by jfmengels** [comment]
> That means it can't flag an issue being moved around, or code being rewritten but reintroducing the exact same issue it removed, right?

Correct. Though it can flag an issue if it's moved around from one file to another, just not in the same file. Also, `elm-review` updates the suppression files silently when it's being run (to remove the pain point of having to re-run the linter to update those files, we're trying to encourage good behavior not punish it), so if you run remove a problem, run `elm-review` (which updates the suppression files), reintroduce an error and finally run `elm-review` again, then it will actually catch the problem. 

As you say, you can basically swap one error in one file for another. While it's not necessarily always true (but generally is), this is fine as long as the severity of all the problems reported by a rule are the same, because you're not making the code necessarily worse, you're keeping it as it was before. And this is the worst scenario with this system, which IMO is a pretty good tradeoff considering the pros and cons of alternative solutions.

> Although what happens if a new issue is detected, does elm-review just output all the issues of the file?

Exactly, it outputs all the issues of the file for that specific rule. This is what the output looks like (focus on the orange parts in this example):
![](https://jfmengels.net/static/e14611ccb5e17b7843300f55b75bba3a/64d87/output-when-more-errors.png)

**Comment by jfmengels** [comment]
For more information on this, I opened a similar issue in ESLint very recently: https://github.com/eslint/eslint/issues/16755

Another prior art is [`ember-template-lint's TODO system`](https://github.com/ember-template-lint/ember-template-lint/blob/master/docs/todos.md). A few tradeoffs that they chose went for:
- There is a separate file for every suppressed error, with a hash of the contents in the file name. This is great for avoiding Git conflicts (which happen more in my system with a single file)
- The file contains some information about the error, including the position of the error in the file
- The date of creation is in the file's contents, which allows them to "snooze" errors after a set amount of time.

Reasons why I didn't go for some of these (that system is older than what `elm-review` did):
- Snoozing errors is IMO not a great system. Errors will show up after a set amount of time, and fail unrelated builds. If I come back to a project after6 months, I don't want to have fix all the errors that now get reported. Instead `elm-review` have a small reminder like "I found no errors, but there are still 12 suppressed errors to address".
- Since the file contains the position, everytime the position gets changed because of unrelated changes above it, the file needs to be regenerated (and I imagine developers just do that constantly without thinking too much? I can't see that not happening)
- Since the files get regenerated all the time because of the previous bullet point, the date of creation also gets changed every time. Meaning that in often changed files, you'll never get reminders for the errors in those files. Snoozing and re-creating these files with hashes are IMO features that don't work well together.
- I like the conflict-less parts, but I went for a single file, because it makes it easy to have an overview of the suppressed errors (which rules are suppressed, how many errors are suppressed) without having to execute the file. Having plenty of files didn't give a good overview IMO.

**Comment by jfmengels** [comment]
This is an example of a `.lint-todo` folder that I could find: https://github.com/dfreeman/ember-truncate/tree/786b812d4e7993c5ba253f5af8a984ed40976291/.lint-todo

I find that neither the file structure neither the file's contents are easy to get an overview.
In comparison, this is what you get for `elm-review`: https://github.com/jfmengels/elm-spa-example/tree/46591cb15b55136afed1a25fe1a5563545a1b3de/review/suppressed

Here, I don't have to run any tool, know any bash-fu, or even leave GitHub to see what's happening. In this last example, I can clearly see which rules have been suppressed, and then I can look at those files to see how many suppressed errors there are.

If I review a GitHub PR, it is easy to see whether new rules are being suppressed or more errors are being suppressed, which allows me to leave a comment asking my colleague whether this was necessary, and start a conversation. We ideally want those numbers to go down, not up, so talking it out is a healthy practice IMO.

With `ember-template-lint`'s approach, I might see 10 removed files and +/- 10 added files, and I'm going to need to dig into all of those to see whether things got better, worse or (because they're location-based and often re-generate) things stayed the same.

> support both styles

I don't think that having more choices for these things is beneficial in general, as that will make it harder to maintain things, confuse the users and split the community in how you do things. And you might end up with the worst of both worlds instead of the best of both worlds.

> by accepting both baseline file and baseline directory

Can you clarify what you have in mind with baseline file and baseline directory?

**Comment by judson-stevens-teampay** [comment]
I'm new to `ruff` (loving it so far!) but have been using `flakeheaven` and it's predecessor `flakehell` for quite a while now. I really enjoy `flakeheaven`'s baseline feature - there's some more information [here](https://github.com/flakeheaven/flakeheaven/blob/main/docs/commands/baseline.md). We've used it successfully on a larger code base to incrementally fix legacy issues by incorporating the `diff` command in that linked doc. We have a fairly large Python monolith at my day job, and when developers make changes they are prompted to fix issues in code they touched; an easy example of this is docstrings - in functions that do not have docstrings, if a developer modifies that function, `flakeheaven` will then identify it as needing a docstring and flag it in `pre-commit` or our CI. 

I'm definitely not an expert in the mechanism used to generate the baseline file, but just wanted to toss in some lived experience with this functionality. I believe it is similar to what others have called out above, in that it's [tied to the filename and the line numbers](https://github.com/flakeheaven/flakeheaven/blob/main/flakeheaven/_logic/_baseline.py), so we do occasionally see instances where very old legacy code is touched and multiple "unrelated" issues then pop up. However, in general, even on a code base with some fairly large legacy files (7k lines of Django serializers), it's usually not so bad to get things fixed up. This incrementally modernizes and even fixes our existing code, which is great.

**Comment by lengau** [comment]
One possible implementation of baselines that would handle many of the issues herein (though it would introduce at least one of its own) would be to incorporate the baseline into the file itself with the `noqa` directive. If `--add-noqa` could include extra information in the comment containing the directive, it could be used to set a baseline. Combined with rule `RUF100` and ruff's auto-fix feature, this could provide a very user-visible version of baselines directly in the code.

For example, a file that contains an existing line of:

    from math import *

could get a directive such as:

    from math import *  # noqa: F403 - Existing errors as of 2023-01-29 ignored by ruff.

In my view it has the following advantages:

1. No messing with line numbers, counting violations, etc.
2. The directives are very visible to the user, encouraging their fixing when touching nearby code
3. A baseline can follow a particular code snippet even if files are moved, etc. Splitting a thousand-line legacy file in two doesn't have to come with the work of fixing all the linting errors in half that file. (After all, a linting tool is supposed to help us, not burden us.)
4. With `--fix` and rule `RUF100`, directives get automatically removed when appropriate.

Of course, it also has one distinct disadvantage:

1. Every rule-violating line of code now has an extra node in the blame layer just for adding `noqa` directives.

This could be mitigated by providing an obvious separate author credit on this (commit the output of `--add-noqa` "as ruff", if you will). Developers would immediately recognise this and look at previous commits, and if a standard is generated tools might begin to provide similar options.

**Comment by orsinium** [comment]
I also have a baseline implementation for mypy as a separate tool: [mypy-baseline](https://github.com/orsinium-labs/mypy-baseline). As people already said in the thread, baseline doesn't store line numbers. In case of mypy-baseline, violations are matched to the baseline based on the file name and error message:
https://github.com/orsinium-labs/mypy-baseline/blob/master/mypy_baseline/_error.py#L60-L68

In the case of flakehell (now flakeheaven) things were a bit easier because I could use error codes (since error messages may change):
https://github.com/life4/flakehell/blob/master/flakehell/_logic/_baseline.py#L6-L11

In case of a new violation being similar to the old one (the same file and the same error code), simply the last one is reported based on assumption that the new code is more often added closer to the end of the file. It may produce false-positives in theory but my teams haven't faced it yet.

I think it was a good decision to keep baselines in a separate file. I also considered automatically adding `# noqa` to the code but abandoned that idea:

1. All `# noqa` (or `typing: ignore` in case of mypy) comments should be used in rare occasions of legit false-poisitives when a human looked at the code and said: "yep, that would be hard to work around". The baseline, on the other hand, is for the tech debt that humans haven't looked into yet but should gradually go through and get rid of. I don't think these two should be mixed.
2. Having the baseline in one place gives a better perspective on the work left to do. In mypy-baseline, I provide `history`, `plot`, and `top-files` commands that I often run and show the team to remind them that we should do progress resolving it.
3. Introducing a tool that just needs a single baseline file and a single config is much easier than a tool that changes something in every file.

What I regret doing in flakehell is that the baseline is a collection of obscure md5 hashes. In mypy-baseline, the baseline is human-readable and we quite often use that fact to manually read and edit it.

I hope this perspective can be of any help.

**Comment by snmishra** [comment]
> That's interesting! I've never seen this before. I think we could support it? I'd been hoping that `--add-noqa` would serve this purpose. What do you see as the strengths of a baseline file vis-a-vis that workflow?
> 
> One point of confusion for me w/r/t the baseline file is that if it's just storing errors + line numbers, then if you edit a file (e.g., add a function to the top of a file that has some ignored errors), doesn't the baseline get invalidated? Since all the errors are pushed to new lines?

flakeheaven [stores md5 hash](https://github.com/flakeheaven/flakeheaven/blob/4591fd3dd04c8180d702e5eab3ec25947d7d3393/flakeheaven/_logic/_baseline.py) of the violation and it's context. If performance is a concern a cheaper hash should be fine too. 

The `--add-noqa` wouldn't leave a way to find out if the `noqa` were never meant to be fixed, or were just added for initial adoption.

**Comment by Pixel-Minions** [comment]
Hi,

I wholeheartedly support the implementation of a separate baseline file as it proves to be an immensely valuable feature. Given the context of my position as the head of the department in a company with a substantial monolithic codebase, I am actively striving to enforce robust code standards. To achieve this, I have introduced linters during the code push to the repository and have provided the team with suitable tools. However, due to the sheer size of the codebase, it is not feasible to lint the entirety of it. Therefore, my objective is to exclusively lint the new code changes—the actual differences in the codebase as it progresses forward.

Flakeheaven or flakehell do a great job doing this baseline, I wish ruff could bring something similar.

**Comment by dorschw** [comment]
As (temporary?) solution, I just released Riff [PyPi](https://pypi.org/project/riff) / [GitHub](https://github.com/dorschw/riff), 

Riff wraps Ruff, runs git diff, checks for modified lines, only fails if errors were found in modified lines.
Works as a CLI tool as well as a pre-commit hook, check it out :)

**Comment by KotlinIsland** [comment]
[Darker](https://github.com/akaihola/darker) is a very relevant tool here.

And for prior art, [basedmypy](https://github.com/KotlinIsland/basedmypy) and [Qodana](https://www.jetbrains.com/qodana/) both support baseline functionality.

**Comment by carlfischerjba** [comment]
Here's a bit of `jq` that will extract the differences between two GitLab-style reports and print it out as a Markdown snippet. Disclaimer: I'm no jq expert so there are probably better ways to do this.

```sh
jq -nr --argfile ref gl-code-quality-report-baseline.json \
    --argfile new gl-code-quality-report.json '
# Get fingerprints in new report that are not in the old.
[ $ref, $new ]
| map([.[].fingerprint])
| (.[1] - .[0]) as $new_fingerprints
# Take the new report and select only the entries with a new fingerprint.
| $new
| .[] | select(.fingerprint | IN($new_fingerprints[]))
| "- \(.description) - `\(.location.path):\(.location.lines.begin)`"
'
```

I've added this in one of our GitLab projects while waiting for [this issue](https://gitlab.com/gitlab-org/gitlab/-/issues/34982) to be resolved. Here's the merge request workflow.
1. Run ruff to create a report.
2. Download the baseline via the GitLab API.
3. Use jq to extract issues that were not present in the baseline.
4. Fail the pipeline if new issues were found.

**Comment by blablatdinov** [comment]
You can use `ondivi` [Pypi](https://pypi.org/project/ondivi/) [Github](https://github.com/blablatdinov/ondivi). Is a Python script filtering coding violations, identified by static analysis, only for changed lines in a Git repo. You can use it with ruff, flake8, pylint, mypy and others.

flakeheaven and flakehell are not supported because they rely on internal flake8 API, which can lead to compatibility issues as flake8 evolves. In contrast, ondivi uses only the text output of violations and the state of Git repository, making it more robust and easier to maintain.

> Flakeheaven: https://wemake-python-styleguide.readthedocs.io/en/latest/pages/usage/integrations/flakeheaven.html#legacy-first-1

WPS remove this page from the docs and replace it with https://wemake-python-styleguide.readthedocs.io/en/latest/pages/usage/integrations/ondivi.html

**Comment by jedie** [comment]
Think this issue is only about to run ruff **linting** only on change lines, but not to run the formatter only on change lines, isn't it?

The mentioned solutions like `ondivi` or `Riff` doesn't help here. `ondivi` is only for linting, and `riff --fix` will apply formatting to all lines, see: https://github.com/dorschw/riff/issues/16

So i end up to wait for https://github.com/akaihola/darker/issues/521 for support "Support ruff format as a formatter" in darker.

**Comment by KotlinIsland** [comment]
basedpyright also supports baseline functionality https://docs.basedpyright.com/latest/benefits-over-pyright/baseline/

**Comment by estahn** [comment]
This is received positive feedback in terms of functionality. Is the only open question the format of the baseline data?

I was imagining this could be stored by hashing the filename with line of code and possibly error code, e.g. `MD5(CONCAT(<filename>, <line of code>, <ruff error code>))`.

**Example:**

Finding:
`test_foobar.py`
```
55 | def test_do_FooBar(api_client, authenticated_user, mock_pynamodb, mock_eventbridge_client):
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ N802
```

```bash
ruff . --exit-zero --format baseline > .ruff_baseline
ruff . --baseline .ruff_baseline  # doesn't report any errors
```

.ruff_baseline
```
<filename>|<md5 of line 55>|<error code>
2af918dbc4589ec4a153761e4c9af86f
```

**Comment by josiahcoad** [comment]
Very excited for this. Is currently needed on my team.

**Comment by InsidersByte** [comment]
[ESLint added support for bulk suppression a few months ago](https://eslint.org/blog/2025/04/introducing-bulk-suppressions/). 

Their RFC for it can be found [here](https://github.com/eslint/rfcs/tree/main/designs/2024-baseline-support).

**Comment by phiresky** [comment]
Until this is implemented here's a one-line command that generates the `[tool.ruff.lint.per-file-ignores]` section based on the currently violated rules:

```bash
uv run ruff check --output-format=json | jq -r 'group_by(.filename) | map({(.[0].filename): ([.[].code] | unique)}) | add | to_entries | map("\"" + .key + "\" = [" + (.value | map("\"" + . + "\"") | join(", ")) + "]") | join("\n")'
```

this generates an output like this:
```
"__init__.py" = ["E402"]
"path/to/file.py" = ["E402", "TRY300"]
```

which can be directly added to the pyproject.toml file under the `[tool.ruff.lint.per-file-ignores]`

At least one rule needs to be failing otherwise jq will throw a null error.

**Comment by epic-64** [comment]
Baseline support is practically a requirement for introducing linters to an existing codebase, especially in teams that are not used to static analysis. The alternative would be to allow linting to fail in CI, or use extremely watered-down rules that do not protect the codebase from new ill-structured code.

I use baselines very successfully in phpstan and mypy (granted, mypy-baseline is a third party package that allows you to pipe mypy output through it - I actually kinda like this separation of concerns).

Looking very much forward to this for Ruff.

**Comment by Artalus** [comment]
PVS-Studio, a paywalled but quite capable C++ static analyzer, [supports](https://pvs-studio.com/en/docs/manual/0032/) generating baseline files. IIRC it can store them in custom paths that can be configured with `--some-cli-arg` when running the analyzer from terminal or in CI.

Their baseline does something like stripping whitespace from the code line – plus previous line, plus next line – and hashing the result. This helps to ensure that re-indenting the problematic fragment and adding new code around does not immediately cause the baselined diagnostics to reappear.

**Comment by siscia** [comment]
As I mentioned in #21907 and @ntBre described better than me.

A reasonable substitute for baseline would it be to allow something like RUF100 to unused `per-file-ignores` 

When introducing a new lint, we would just add the files to the `per-file-ignores` and as soon as the lint is addressed the files can be removed by the `per-file-ingores` list.

The file could accumulate more of those exceptions, but I don't see it as a practical issue.

**Comment by edgarrmondragon** [comment]
@siscia that'd be a nice improvement but a baseline is usually more granular than simply ignoring an entire file, like what @Artalus described

**Comment by siscia** [comment]
Hello,

just for my understanding here. What's the status? 

Is astral on board with baseline being implemented and added to ruff?
Is astral on board with RUF100 to be available also for `per-file-ignores` being implemented and added to ruff?

Are we still discussing on design? 

Are we still deciding if the feature is worth having?

I don't have capacity myself to provide code in the ruff codebase (sorry about that) but I am more than open to provide examples of how we will use either features in our codebase. Provide feedback on possible designs from a user perspective. And pretty much whatever would help without require me 2+ hours of deep work on this.

If there is any way, I can help move this forward, please let me know.

Thanks a lot

**Comment by MichaReiser** [comment]
It's on our mind, but not a top-priority feature. We plan to come back to it, but I can't say yet when.

**Comment by ofekby** [comment]
In the meantime, try using snaplint to snapshot your current linter output and only fail CI on new issues — it’s a simple way to track lint regressions until baseline support lands in Ruff: [https://github.com/GENWAY-AI/snaplint](https://github.com/GENWAY-AI/snaplint?utm_source=chatgpt.com)

Usage example:
1. Run once and push snapshot files 
```
ruff check --output-format concise | snaplint take-snapshot
```

2. Then on every CI execution
```
ruff check --output-format concise | snaplint diff
```

**Comment by dylwil3** [comment]
Apologies for the notification spam to all watching this issue! 

Could folks indicate (preferably with emoji reactions to this comment, to prevent further notification spamming) whether

```
ruff check --add-noqa="# Baseline"
```

would address most of your use-cases? 

This is almost identical to Charlie's original suggestion, except that in [Ruff 0.14.5](https://github.com/astral-sh/ruff/releases/tag/0.14.5) Micha added the option to give a "reason" that appears next to each added `noqa`. 

For example, if you had only `F401` enabled and the file `ex.py` with

```python
import os
```
you would get:

```python
import os  # noqa: F401 # Baseline
```

That's greppable and distinguishes the noqa from an "intentional" noqa that didn't come from a baseline. This would also avoid any of the shortcomings (both in design and implementation) of a file that "remembers" what errors existed in some fixed version of your code (and yet behaves correctly as the code evolves).

It does not handle the case of the formatter - so let's say I'm only asking about the linter for the purposes of this "poll".

Edit: Oh dear, I apparently missed where this was [suggested earlier in this thread](https://github.com/astral-sh/ruff/issues/1149#issuecomment-1407937056) (prior to the feature being implemented)! But in case my even-spammier-than-intended-spam brought you here, it'd still be nice to get a thumbs up/down to see whether "add noqa with reason" is an acceptable solution to this problem.

Incidentally, the concern about `git blame` could be mitigated with [`--ignore-revs-file`](https://git-scm.com/docs/git-blame#Documentation/git-blame.txt---ignore-revs-filefile).

**Comment by cryptomilk** [comment]
> ```
> ruff check --add-noqa="# Baseline"
> ```
>
> [...]
>
> you would get:
> 
> import os  # noqa: F401 # Baseline

I'm answering here as a Samba Developer. This would be unacceptable by the Samba project and probably by a lot of other projects too. It would change the code and thus change the git history. It means `git annotate` stops working. Samba is a huge complex code base and we rely on the working git history on a daily base.

That's the reason why basedpyright created a separate file you can turn on and of at will :-)

**Comment by Avasam** [comment]
Edit: sorry, I missed this was already mentioned: https://github.com/astral-sh/ruff/issues/1149#issuecomment-3137819224

> [ESLint added support for bulk suppression a few months ago](https://eslint.org/blog/2025/04/introducing-bulk-suppressions/).
> 
> Their RFC for it can be found [here](https://github.com/eslint/rfcs/tree/main/designs/2024-baseline-support?rgh-link-date=2025-07-30T21%3A06%3A22.000Z).

---

ESLint introduced this through their new "Bulk Suppression" feature: https://eslint.org/docs/latest/use/suppressions

There's a file that lists every error code and their count for every file. It's only tracking count, not position. And honestly that's enough to check if "new" errors are introduced. (technically false-positives if you fix one and introduce the same error code elsewhere in the same file, but that really doesn't matter much).

Let's say in `my_module.py` you have 3 `ISC001` errors. and in `ruff-suppressions.toml` the file `my_module.py` is marked as currently having 2 `ISC001` errors, then only the 3rd one would be raised by Ruff.

In ESLint's implementation, you can easily generate AND cleanup the suppression file through a command line. So you can generate it once initially, then just add the cleanup flag to your existing "fix" command.

**Comment by paulburlumi** [comment]
golangci-lint has `new-from-rev`, which only reports issues on lines added
after a given revision. It's what they recommend for adopting the linter on an
existing project that has a lot of violations:
https://golangci-lint.run/docs/welcome/faq/#how-to-integrate-golangci-lint-into-a-large-project-with-thousands-of-issues

The useful bit is that the baseline is just a commit SHA in the config file:

```yaml
issues:
  new-from-rev: 08f836be69dc11522479505d295c9889d9268042
```

So there's no generated baseline to keep up to date or resolve conflicts in,
and you move the line forward by editing one string. `--new-from-merge-base=main`
does the same thing against a moving base if you want it per-PR instead, and
`--whole-files` widens it to every issue in a touched file when line-level
matching is too strict.

That seems like a lot less work than a serialized baseline format, and it would
cover what most people in this thread are asking for.

**Comment by MichaReiser** [comment]
Thanks, this is interesting. I'm not sure if this is necessarily simpler. I mean, it is for users. I'm not sure it is simpler to implement. I'd have to look at what go does in this case but I imagine go runs twice, once on the baseline ref and once on the latest ref and then does some diffing to identify new diagnostics.

**Comment by paulburlumi** [comment]
> I'd have to look at what go does in this case but I imagine go runs twice, once on the baseline ref and once on the latest ref and then does some diffing to identify new diagnostics.

I was interested too, so I had Claude analyse the repo.

No - it doesn't run the linters twice. It's much simpler (and cheaper): lint once on the working tree, then filter the results by a git diff.

1. Get a patch. With `new-from-rev=X` it shells out to `git diff --color=never --no-ext-diff --default-prefix --relative X --` and also runs `git ls-files --others --exclude-standard` so untracked files count as entirely-new. (`new-from-merge-base=X` first resolves `git merge-base X HEAD` and uses that as the from-revision; bare `new` uses unstaged changes, falling back to `HEAD~`.)
1. Parse the patch into changed line numbers. It scans the unified diff, tracking `+++ b/<file>` headers and `@@` hunk headers, and builds `map[filePath][]pos` where each `pos` is `{lineNo, hunkPos}` for every added (`+`) line. Untracked files map to a `nil` slice, meaning "whole file is new".
1. Filter diagnostics. For each issue, it looks up the file and checks whether the issue's line number is among the added lines. Not in the map → dropped. In the map → kept. With `whole-files: true`, any issue anywhere in a touched file is kept.

Consequences of this design

- Fast — one lint run, one git diff. No baseline checkout, no second analysis pass.

**Comment by MichaReiser** [comment]
I see, thanks. So touching a line will always report it as new (or reformatting). I do like the simplicity
